src/rules/has-valid-accessibility-state.js

```diff
@@ -1,40 +1,60 @@
 /**
- * @fileoverview Used to tell Talkback or Voiceover the state a UI Element is in
- * @author Jen Luker
+ * @fileoverview Describes the current state of a component to the user of an assistive technology.
+ * @author JP Driver
  * @flow
  */
 
-import createValidPropRule from '../factory/valid-prop';
+import type { JSXAttribute } from 'ast-types-flow';
+import { elementType, getPropValue, getLiteralPropValue } from 'jsx-ast-utils';
+import { generateObjSchema } from '../util/schemas';
+import type { ESLintContext } from '../../flow/eslint';
 
 // ----------------------------------------------------------------------------
 // Rule Definition
 // ----------------------------------------------------------------------------
 
-const errorMessage =
-  'accessibilityStates must be one or both of the defined values';
+const validKeys = ['disabled', 'selected', 'checked', 'busy', 'expanded'];
 
-const validValues = ['selected', 'disabled'];
+module.exports = {
+  meta: {
+    docs: {},
+    schema: [generateObjSchema()]
+  },
 
-let deprecationHasBeenWarned = false;
+  create: (context: ESLintContext) => ({
+    JSXAttribute: (node: JSXAttribute) => {
+      const attrName = elementType(node);
+      if (attrName === 'accessibilityState') {
+        const attrValue = getPropValue(node);
+        const test = getLiteralPropValue(node);
 
-const rule = createValidPropRule(
-  'accessibilityStates',
-  validValues,
-  errorMessage,
-  {
-    deprecated: true
-  },
-  {
-    Program: () => {
-      if (deprecationHasBeenWarned) return;
-      // eslint-disable-next-line no-console
-      console.log(
-        'The react-native-a11y/has-valid-accessibility-state rule is deprecated. ' +
-          'Please use the react-native-a11y/has-valid-accessibility-states rule instead.'
-      );
-      deprecationHasBeenWarned = true;
-    }
-  }
-);
+        const error = message =>
+          context.report({
+            node,
+            message
+          });
 
-module.exports = rule;
+        if (typeof attrValue !== 'object' || Array.isArray(attrValue)) {
+          error('accessibilityState must be an object');
+        } else {
+          Object.entries(attrValue).map(([key, value]) => {
+            if (validKeys.indexOf(key) < 0) {
+              error(`accessibilityState object: "${key}" is not a valid key`);
+            } else if (key !== 'checked' && typeof value !== 'boolean') {
+              error(
+                `accessibilityState object: "${key}" value is not a boolean`
+              );
+            } else if (
+              key === 'checked' &&
```

Comment:
Nitpick, but would the logic not look a (tiny) bit nicer if we put this `else if` above the one on line `43`, removing the need for `key !== 'checked` there?

Reply:
I know the logic is a bit verbose here -- honestly that's mainly for myself so I don't have to think too much to re-grok what the rule is doing next time I'm working on this.

I'm happy to move this up and swap them round; but we do need both conditions here, otherwise lines like this:
```
<TouchableOpacity accessibilityState={{ checked: "mixed" }} />
```
would fail.

We'd first check that `checked` is either a bool or "mixed" -- which it is, so no error -- but then if we get rid of the second `key !== 'checked'` the final condition is only `typeof value !== 'boolean'` which this case fails.

We can definitely re-arrange if you prefer, but because they're inter-connected we do need both checks 🙂